### ISSUE TITLE
Unable to deserialize custom RNN cell decorated via register_keras_serializable when used inside tf.keras.layers.RNN

### ISSUE THREAD

**Issue by mlech26l** [tensorflow/tensorflow#45224]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- Ubuntu 18.04:
- TensorFlow 2.3.1
- Python 3.6:
- CUDA 10.2, cuDNN 7.6.5.32
- GeForce GTX 1080 Ti

**Describe the current behavior**

Custom RNNCell layer decorated by ```tf.keras.utils.register_keras_serializable``` are not found when deserializing a ```tf.keras.layers.RNN``` object via ```tf.keras.models.load_model```

**Describe the expected behavior**
Custom layers decorated by ```register_keras_serializable``` should be able to be deserialized when used inside ```tf.keras.layers.RNN```. 

**Standalone code to reproduce the issue**
```python
import numpy as np
import tensorflow as tf

data_x = np.random.default_rng().normal(size=(100, 16, 10))
data_y = np.random.default_rng().normal(size=(100, 16, 1))

@tf.keras.utils.register_keras_serializable(package="Custom")
class CustomLayer(tf.keras.layers.Layer):
    def __init__(self, units, outputs, **kwargs):
        self.units = units
        self.outputs = outputs
        super(CustomLayer, self).__init__(**kwargs)

    @property
    def state_size(self):
        return self.units

    def build(self, input_shape):
        self.w = self.add_weight("w", shape=(input_shape[-1], self.units))
        self.o = self.add_weight("o", shape=(self.units, self.outputs))
        self.r = self.add_weight("r", shape=(self.units, self.units))
        self.built = True

    def call(self, inputs, states):
        next_hidden = tf.nn.tanh(
            tf.matmul(inputs, self.w) + tf.matmul(states[0], self.r)
        )
        output = tf.matmul(next_hidden, self.o)
        return output, [next_hidden]

    def get_config(self):
        return {"units": self.units, "outputs": self.outputs}

model = tf.keras.models.Sequential(
    [
        tf.keras.Input((16, 10,)),
        tf.keras.layers.RNN(CustomLayer(10, 1), return_sequences=True),
    ]
)
model.compile(optimizer="adam", loss="mean_squared_error")

model.fit(x=data_x, y=data_y, batch_size=25, epochs=1)
model.evaluate(x=data_x, y=data_y)
model.save("test_2.h5")

model = tf.keras.models.load_model("test_2.h5")  ## Crashes here
model.evaluate(x=data_x, y=data_y)
```

**Comment by amahendrakar** [comment]
@mlech26l,
On adding the argument `custom_objects={"CustomLayer": CustomLayer}` to the `load_model` function I was able to run the code without any issues. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/0b9c130c321af03470c7f234ce4fd068/45224.ipynb). 

For more information, please go through [this guide](https://www.tensorflow.org/guide/keras/save_and_serialize#custom_objects) for saving models with custom objects and [this similar](https://stackoverflow.com/a/50838031) StackOverflow query. Thanks!

**Comment by mlech26l** [comment]
@amahendrakar thanks. I was already aware of the ```custom_objects``` argument workaround, however, it does not provide a clean solution for third-party libraries that implement several RNN cell classes. 

Thus, will this feature be added in the future or do third-party libraries have to provide a ```dict``` storing all serializable RNN cell?
Thanks!

**Comment by jvishnuvardhan** [comment]
@mlech26l Can you please share the code where `load_model` fails with third-party libraries? Thanks!

**Comment by mlech26l** [comment]
Hi @jvishnuvardhan 
The [keras-ncp](https://github.com/mlech26l/keras-ncp) define RNNCells similar to the snippet above and crashes. 
Currently, the only option is to pass the ```custom_objects``` parameter to the ```load_model``` function, which is a sub-optimal solution. 
I highly expect that the Keras-API decorator is intended to cover the use inside an RNN object as well.

Thanks.

**Comment by mlech26l** [comment]
Hi, 
This issue has been fixed with the 2.4 release

Thank you
